File: src/main/java/org/radarbase/appserver/service/QuestionnaireScheduleService.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.service;

import lombok.extern.slf4j.Slf4j;
import org.radarbase.appserver.dto.protocol.Assessment;
import org.radarbase.appserver.dto.protocol.AssessmentType;
import org.radarbase.appserver.dto.protocol.Protocol;
import org.radarbase.appserver.dto.questionnaire.AssessmentSchedule;
import org.radarbase.appserver.dto.questionnaire.Schedule;
import org.radarbase.appserver.entity.Project;
import org.radarbase.appserver.entity.Task;
import org.radarbase.appserver.entity.User;
import org.radarbase.appserver.exception.NotFoundException;
import org.radarbase.appserver.repository.ProjectRepository;
import org.radarbase.appserver.repository.UserRepository;
import org.radarbase.appserver.search.TaskSpecificationsBuilder;
import org.radarbase.appserver.service.questionnaire.protocol.ProtocolGenerator;
import org.radarbase.appserver.service.questionnaire.schedule.QuestionnaireScheduleGeneratorService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Slf4j
@Service
@Scope(value = ConfigurableBeanFactory.SCOPE_SINGLETON)
@SuppressWarnings("PMD.ConstructorCallsOverridableMethod")
public class QuestionnaireScheduleService {

    private static final Pattern TASK_SEARCH_PATTERN = Pattern.compile("(\\w+)([:<>])(\\w+)");
    private static final Pattern COMMA_PATTERN = Pattern.compile(",");

    private final transient ProtocolGenerator protocolGenerator;

    private final transient HashMap<String, Schedule> subjectScheduleMap = new HashMap<>();

    private final transient UserRepository userRepository;

    private final transient QuestionnaireScheduleGeneratorService scheduleGeneratorService;

    private final transient ProjectRepository projectRepository;

    @Autowired
    private final transient TaskService taskService;

    @Autowired
    private final transient FcmNotificationService notificationService;

    @Autowired
    public QuestionnaireScheduleService(ProtocolGenerator protocolGenerator, UserRepository userRepository, ProjectRepository projectRepository, QuestionnaireScheduleGeneratorService scheduleGeneratorService, TaskService taskService, FcmNotificationService notificationService) {
        this.userRepository = userRepository;
        this.projectRepository = projectRepository;
        this.taskService = taskService;
        this.protocolGenerator = protocolGenerator;
        this.scheduleGeneratorService = scheduleGeneratorService;
        this.notificationService = notificationService;
    }


    @Transactional
    public List<Task> getTasksUsingProjectIdAndSubjectId(String projectId, String subjectId) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        return this.getTasksForUser(user);
    }

    @Transactional
    public List<Task> getTasksByTypeUsingProjectIdAndSubjectId(String projectId,
                                                               String subjectId,
                                                               AssessmentType type,
                                                               String search) {

        Specification<Task> spec = getSearchBuilder(projectId, subjectId, type, search).build();
        return this.taskService.getTasksBySpecification(spec);

//        if (type != AssessmentType.ALL) {
//            return getTasksUsingProjectIdAndSubjectId(projectId, subjectId);
//        } else {
//            User user = subjectAndProjectExistElseThrow(subjectId, projectId);
//            return this.taskRepository.findByUserIdAndType(user.getId(), type);
//        }
    }

    @Transactional
    public List<Task> getTasksForDateUsingProjectIdAndSubjectId(String projectId, String subjectId, Instant startTime, Instant endTime) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        List<Task> tasks = this.getTasksForUser(user);
        tasks.removeIf(t-> {
            Instant timestamp = t.getTimestamp().toInstant();
            return timestamp.plusMillis(t.getCompletionWindow()).isBefore(startTime) || timestamp.isAfter(endTime);
        });
        return tasks;
    }

    @Transactional
    public List<Task> getTasksForUser(User user) {
        return this.taskService.getTasksByUser(user);
    }

    @Transactional
    public Schedule generateScheduleUsingProjectIdAndSubjectId(String projectId, String subjectId) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        return this.generateScheduleForUser(user);
    }

    @Transactional
    public Schedule generateScheduleForUser(User user) {
        Protocol protocol = protocolGenerator.getProtocolForSubject(user.getSubjectId());
        Schedule newSchedule;

        if (protocol == null) {
            newSchedule = new Schedule();
        } else {
            Schedule prevSchedule = getScheduleForSubject(user.getSubjectId());
            String prevTimezone = prevSchedule.getTimezone() != null
                    ? prevSchedule.getTimezone()
                    : user.getTimezone();
            if (!Objects.equals(prevSchedule.getVersion(), protocol.getVersion()) || !prevTimezone.equals(user.getTimezone())) {
                this.removeScheduleForUser(user);
            }
            newSchedule = this.scheduleGeneratorService.generateScheduleForUser(user, protocol, prevSchedule);
        }

        subjectScheduleMap.put(user.getSubjectId(), newSchedule);
        this.saveTasksAndNotifications(newSchedule.getAssessmentSchedules(), user);

        return newSchedule;
    }

    private void saveTasksAndNotifications(List<AssessmentSchedule> assessmentSchedules, User user) {
        assessmentSchedules.stream()
                .filter(Objects::nonNull)
                .filter(AssessmentSchedule::hasTasks)
                .forEach(a -> {
                    this.taskService.addTasks(a.getTasks(), user);
                    this.notificationService.addNotifications(a.getNotifications(), user);
                    this.notificationService.addNotifications(a.getReminders(), user);
                });
    }

    @Transactional
    public Schedule generateScheduleUsingProjectIdAndSubjectIdAndAssessment(String projectId,
                                                                            String subjectId,
                                                                            Assessment assessment) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        Protocol protocol = protocolGenerator.getProtocolForSubject(user.getSubjectId());
        if (!protocol.hasAssessment(assessment.getName())) {
            throw new NotFoundException("Assessment not found in protocol. Add assessment to protocol first.");
        }

        Schedule schedule = getScheduleForSubject(user.getSubjectId());
        AssessmentSchedule a = scheduleGeneratorService.generateSingleAssessmentSchedule(
                assessment, user, Collections.emptyList(), user.getTimezone());
        schedule.addAssessmentSchedule(a);

        this.saveTasksAndNotifications(List.of(a), user);
        return schedule;
    }

    @Scheduled(fixedRate = 3_600_000)
    public void generateAllSchedules() {
        List<User> users = this.userRepository.findAll();
        log.info("Generating all schedules..");
        users.forEach(this::generateScheduleForUser);
    }

    public Schedule getScheduleForSubject(String subjectId) {
        Schedule schedule = subjectScheduleMap.get(subjectId);
        return schedule != null ? schedule : new Schedule();
    }

    @Transactional
    public void removeScheduleForUserUsingSubjectIdAndType(String projectId,
                                                           String subjectId,
                                                           AssessmentType type,
                                                           String search) {
        Specification<Task> spec = getSearchBuilder(projectId, subjectId, type, search).build();

        // TODO: DeleteAll with Specifications will soon be released in JPA (v 3.0.0), so update this to not fetch all entities.
        this.taskService.deleteTasksBySpecification(spec);
    }

    @Transactional
    public void removeScheduleForUser(User user) {
        this.taskService.deleteTasksByUserId(user.getId());
    }

    public User subjectAndProjectExistElseThrow(String subjectId, String projectId) {
        Project project = this.projectRepository.findByProjectId(projectId);
        if (project == null) {
            throw new NotFoundException(
                    "Project Id does not exist. Please create a project with the ID first");
        }

        User user =
                this.userRepository.findBySubjectIdAndProjectId(subjectId, project.getId());
        if (user == null) {
            throw new NotFoundException("Subject Id does not exist. Please create a user with the ID first");
        }

        return user;
    }

    private TaskSpecificationsBuilder getSearchBuilder(String projectId,
                                                       String subjectId,
                                                       AssessmentType type,
                                                       String search) {

        User user = subjectAndProjectExistElseThrow(subjectId, projectId);

        TaskSpecificationsBuilder builder = new TaskSpecificationsBuilder();
        builder.with("user", ":", user);

        if (type != AssessmentType.ALL && type != null) {
            builder.with("type", ":", type);
        }

        if (search != null && !search.isBlank()) {
            String[] searchTerms = COMMA_PATTERN.split(search);

            for (String searchTerm : searchTerms) {
                Matcher matcher = TASK_SEARCH_PATTERN.matcher(searchTerm.trim());
                if (matcher.matches()) {
                    builder.with(matcher.group(1), matcher.group(2), matcher.group(3));
                }
            }
        }
        return builder;
    }


}

File: src/main/java/org/radarbase/appserver/service/QuestionnaireScheduleService.kt
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.service

import org.radarbase.appserver.dto.protocol.Assessment
import org.radarbase.appserver.dto.protocol.AssessmentType
import org.radarbase.appserver.dto.protocol.Protocol
import org.radarbase.appserver.dto.questionnaire.AssessmentSchedule
import org.radarbase.appserver.dto.questionnaire.Schedule
import org.radarbase.appserver.entity.Task
import org.radarbase.appserver.entity.User
import org.radarbase.appserver.exception.NotFoundException
import org.radarbase.appserver.repository.ProjectRepository
import org.radarbase.appserver.repository.UserRepository
import org.radarbase.appserver.search.TaskSpecificationsBuilder
import org.radarbase.appserver.service.questionnaire.protocol.ProtocolGenerator
import org.radarbase.appserver.service.questionnaire.schedule.QuestionnaireScheduleGeneratorService
import org.radarbase.appserver.util.checkInvalidDetails
import org.radarbase.appserver.util.checkPresence
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.config.ConfigurableBeanFactory
import org.springframework.context.annotation.Scope
import org.springframework.data.jpa.domain.Specification
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.sql.Timestamp
import java.time.Instant

@Service
@Scope(value = ConfigurableBeanFactory.SCOPE_SINGLETON)
class QuestionnaireScheduleService(
    private val protocolGenerator: ProtocolGenerator,
    private val userRepository: UserRepository,
    private val scheduleGeneratorService: QuestionnaireScheduleGeneratorService,
    private val projectRepository: ProjectRepository,
    private val taskService: TaskService,
    private val notificationService: FcmNotificationService
) {
    private val subjectScheduleMap: MutableMap<String?, Schedule> = hashMapOf()

    @Transactional
    fun getTasksUsingProjectIdAndSubjectId(subjectId: String?, projectId: String?): List<Task> {
        return getTasksForUser(subjectAndProjectExistsElseThrow(subjectId, projectId))
    }

    @Transactional
    fun getTasksByTypeUsingProjectIdAndSubjectId(
        projectId: String?,
        subjectId: String?,
        type: AssessmentType?,
        search: String?
    ): List<Task> {
        getSearchBuilder(projectId, subjectId, type, search).build().also { spec ->
            return this.taskService.getTasksBySpecification(spec)
        }

        /** Not sure about it, Keeping it (as it is) from java to refactored kotlin file, just in case needed in future */
//                if (type != AssessmentType.ALL) {
//            return getTasksUsingProjectIdAndSubjectId(projectId, subjectId);
//        } else {
//            User user = subjectAndProjectExistElseThrow(subjectId, projectId);
//            return this.taskRepository.findByUserIdAndType(user.getId(), type);
//        }
    }

    @Transactional
    fun getTasksForDateUsingProjectIdAndSubjectId(
        subjectId: String?,
        projectId: String?,
        startTime: Instant?,
        endTime: Instant?
    ): List<Task> {
        val user: User = subjectAndProjectExistsElseThrow(subjectId, projectId)
        val tasks: MutableList<Task> = this.getTasksForUser(user).toMutableList()

        tasks.removeIf { task ->
            val taskTime: Timestamp? = task.timestamp
            checkNotNull(taskTime) { "Task timestamp cannot is null in questionnaire scheduler service." }

            val completionWindow: Long? = task.completionWindow
            checkNotNull(completionWindow) { "Task completion window is null in questionnaire scheduler service." }

            taskTime.toInstant().let { taskTimeInstant ->
                taskTimeInstant.plusMillis(completionWindow).isBefore(startTime) || taskTimeInstant.isAfter(endTime)
            }
        }

        return tasks
    }

    @Transactional
    fun generateScheduleUsingProjectIdAndSubjectId(subjectId: String?, projectId: String?): Schedule {
        return subjectAndProjectExistsElseThrow(subjectId, projectId).run {
            generateScheduleForUser(this)
        }
    }

    @Transactional
    fun generateScheduleForUser(user: User): Schedule {
        val subjectId: String? = user.subjectId
        checkNotNull(subjectId) { "Subject ID cannot be null in questionnaire scheduler service." }

        val protocol: Protocol? = protocolGenerator.getProtocolForSubject(subjectId)
        val newSchedule: Schedule
        if (protocol == null) {
            newSchedule = Schedule()
        } else {
            val prevSchedule: Schedule = getScheduleForSubject(subjectId)
            val prevTimeZone: String = prevSchedule.timezone ?: user.timezone!!

            if ((prevSchedule.version != protocol.version) || (prevTimeZone != user.timezone)) {
                removeScheduleForUser(user)
            }
            newSchedule = scheduleGeneratorService.generateScheduleForUser(user, protocol, prevSchedule)
        }
        subjectScheduleMap[user.subjectId] = newSchedule
        saveTasksAndNotifications(user, newSchedule.assessmentSchedules)
        return newSchedule
    }

    fun saveTasksAndNotifications(user: User, assessmentSchedules: List<AssessmentSchedule?>) {
        assessmentSchedules.filterNotNull()
            .filter(AssessmentSchedule::hasTasks)
            .forEach {
                taskService.addTasks(it.tasks, user)
                notificationService.addNotifications(it.notifications, user)
                notificationService.addNotifications(it.reminders, user)
            }
    }

    fun generateScheduleUsingProjectIdAndSubjectIdAndAssessment(
        projectId: String,
        subjectId: String,
        assessment: Assessment
    ): Schedule {

        val user: User = subjectAndProjectExistsElseThrow(subjectId, projectId)
        val protocol: Protocol? = protocolGenerator.getProtocolForSubject(subjectId)

        checkInvalidDetails<NotFoundException>(
            {  protocol == null || !protocol.hasAssessment(assessment.name) },
            { "Assessment not found in protocol. Add assessment to protocol first" }
        )

        val userTimeZone = user.timezone
        checkNotNull(userTimeZone) { "User timezone cannot be null in questionnaire scheduler service." }

        val schedule = getScheduleForSubject(subjectId)
        val assessmentSchedule = scheduleGeneratorService.generateSingleAssessmentSchedule(
            assessment,
            user,
            emptyList(),
            userTimeZone
        )

        schedule.addAssessmentSchedule(assessmentSchedule)

        saveTasksAndNotifications(user, listOf(assessmentSchedule))

        return schedule
    }

    @Scheduled(fixedRate = 3_600_000)
    fun generateAllSchedules() {
        logger.info("Generating all schedules")
        userRepository.findAll().also { users: List<User> ->
            users.forEach(this::generateScheduleForUser)
        }
    }

    fun getScheduleForSubject(subjectId: String): Schedule {
        val schedule: Schedule? = subjectScheduleMap[subjectId]
        return schedule ?: Schedule()
    }

    @Transactional
    fun getTasksForUser(user: User): List<Task> {
        return taskService.getTasksByUser(user)
    }

    fun subjectAndProjectExistsElseThrow(subjectId: String?, projectId: String?): User {
        return checkPresence(this.projectRepository.findByProjectId(projectId)) {
            "Project with projectId $projectId not found. Please create the project first."
        }.let { project ->
            checkPresence(this.userRepository.findBySubjectIdAndProjectId(subjectId, project.id)) {
                "User with subjectId $subjectId not found. Please create the user first."
            }
        }
    }

    fun removeScheduleForUser(user: User) {
        taskService.deleteTasksByUserId(user.id)
    }

    fun removeScheduleForUserUsingSubjectIdAndType(projectId: String, subjectId: String, type: AssessmentType, search: String) {
        val specification: Specification<Task>? = getSearchBuilder(projectId, subjectId, type, search).build()

        taskService.deleteTasksBySpecification(specification)
    }

    fun getSearchBuilder(
        projectId: String?,
        subjectId: String?,
        type: AssessmentType?,
        search: String?
    ): TaskSpecificationsBuilder {
        val builder = TaskSpecificationsBuilder()

        subjectAndProjectExistsElseThrow(subjectId, projectId).also { user ->
            builder.with("user", ":", user)
        }

        if (type != AssessmentType.ALL && type != null) {
            builder.with("type", ":", type)
        }
        if (!search.isNullOrBlank()) {
            search.split(COMMA_PATTERN).forEach { searchTerm: String ->
                TASK_SEARCH_PATTERN.matchEntire(searchTerm.trim())?.also { matcher ->
                    val (field, operator, value) = matcher.destructured
                    builder.with(field, operator, value)
                }
            }
        }
        return builder
    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(QuestionnaireScheduleService::class.java)

        private val TASK_SEARCH_PATTERN = Regex("(\\w+)([:<>])(\\w+)")
        private val COMMA_PATTERN = Regex(",")
    }
}